### ISSUE TITLE
Expose gRPC client's backing dispatcher

### ISSUE THREAD

**Issue by troshko111** [envoyproxy/envoy#6378]
*Title*: Expose gRPC client's backing dispatcher

*Description*:
gRPC clients produced by an `AsyncClientFactory` are thread aware (correct me if i'm wrong, but they seem to be bound to the thread of the thread local cluster manager), 

here's the relevant doc comment:
`envoy/include/envoy/grpc/async_client_manager.h`
```c++
// Per-service factory for Grpc::AsyncClients. This factory is thread aware and will instantiate
// with thread local state. Clients will use ThreadLocal::Instance::dispatcher() for event handling.
```

There however does not seem to be a way to get access to the clients dispatcher, which would be useful to say create a timer running on the same thread as the client's callbacks (so your callbacks and the timer callback are single threaded and do not require any synchronization).

Context: say one is implementing an extension like an http filter, one has access to `Server::Configuration::FactoryContext` which provides a gRPC client factory via
`context.clusterManager().grpcAsyncClientManager().factoryForGrpcService(...)`

As far as I understand, if this is called from the http filter factory itself (on the stack of `createFilterFactoryFromProtoTyped`), it would essentially bind to the main thread dispatcher so in this case one can dispatch custom events using the main thread dispatcher and assume gRPC callbacks would be dispatched through the same one.

This seems a bit inconvinient hence the suggestion to expose the dispatcher reference from the gRPC client itself.

Note, `include/envoy/http/async_client.h` already does this (exposes the backing dispatcher), it'd be useful if `include/envoy/grpc/async_client.h` exposed the backing dispatcher in the same way.

Example:
```c++
class Refresher : public Grpc::TypedAsyncStreamCallbacks<com::example::RefreshRequest> {
public:
  explicit Refresher(Grpc::AsyncClientFactoryPtr);

  void onCreateInitialMetadata(Http::HeaderMap&) override;
  void onReceiveInitialMetadata(Http::HeaderMapPtr&&) override;
  void onReceiveMessage(std::unique_ptr<com::example::RefreshRequest>&&) override;
  void onReceiveTrailingMetadata(Http::HeaderMapPtr&&) override;
  void onRemoteClose(Grpc::Status::GrpcStatus, const std::string&) override;

private:
  void scheduleCallback() const;

  Grpc::AsyncClientPtr client_;
  Grpc::AsyncStream* stream_;

  Event::TimerPtr callback_timer_;
};
```

in this example the refresher would create a gRPC client using the supplied factory, but it won't be able to `scheduleCallback` using a timer bound to the gRPC client's dispatcher since one is not available, in this case a callback would need to be scheduled using the main thread's dispatcher (the only accessible) or the current thread's dispatcher (by allocating a new slot and getting one this way) but neither is guaranteed to be the same as the client's dispatcher so the callback would need to be synchronized.

**Comment by lizan** [comment]
gRPC client is different from HTTP client, which is not always backed by an Envoy dispatcher, for Google gRPC client it is running on a side thread and posted back to each worker thread. In your example, would passing a reference of Dispatcher into the Refresher constructor from the calling side work?

**Comment by mattklein123** [comment]
@troshko111 this is what https://github.com/envoyproxy/envoy/blob/e1450a1ca006f7d8b118a63b0f7e30be2639b881/include/envoy/http/filter.h#L112 is for. Does that do what you need?

**Comment by troshko111** [comment]
> In your example, would passing a reference of Dispatcher into the Refresher constructor from the calling side work?

that's basically what I'm trying to achieve, `Refresher` is a background worker if you will, it needs a gRPC client and a way to schedule callbacks on the same thread as the gRPC client's thread executing gRPC callbacks.
I'm trying to get a single threaded model in the `Refresher`, this is applicable to any code which implements gRPC client callbacks and wants to schedule its own callbacks.

@mattklein123 unfortunately not exactly, in my example `Refresher` is a singleton created by the http filter factory on the first invocation so the relationship is single `Refresher` ->* Multiple http filters. Hence I cannot just grab the filter's dispatcher, there's no single one.

A recap:

an http filter factory is creating custom filters, they all get a shared ptr to a singleton `Refresher` which is created the first time said factory creates an http filter.

`Refresher` uses gRPC client to make outbound calls. It needs to schedule some timer based callbacks and in order to avoid synchronization between these and the gRPC callbacks interface of which `Refresher` implements, `Refresher` needs to have a reference to the dispatcher used to dispatch gRPC callbacks.

**Comment by lizan** [comment]
@troshko111 In that case, "Expose gRPC client's backing dispatcher" doesn't seem to achieve that, what you need IIUC is thread local slot? See https://github.com/envoyproxy/envoy/blob/master/include/envoy/thread_local/thread_local.h

**Comment by troshko111** [comment]
hmm why not, I only instantiate 1 gRPC client and pass it to the singleton's ctor? What am I missing in this scenario?

**Comment by lizan** [comment]
> an http filter factory is creating custom filters, they all get a shared ptr to a singleton Refresher which is created the first time said factory creates an http filter.

@troshko111 you have multiple http filters associated to the `Refresher` so they can be in different thread (i.e. different dispatchers), right? You either need to pass those dispatchers individually (via another method) or use thread local thread, no?

**Comment by troshko111** [comment]
they all have a shared_ptr to the same singleton `Refresher`, they can be in different threads but that's irrelevant, for simplicity let's assume they do not call any methods on `Refresher`, only hold a pointer to it.

They hold a pointer to some object doing background processing.

**Comment by lizan** [comment]
Then where are you trying to schedule callback to? In that case you can only schedule back to main thread dispatcher if those filters doesn't call any method and the async client is created by factory (which is in main thread context)

**Comment by troshko111** [comment]
Yes I mentioned this in the initial issue description, it is in fact using the main thread's dispatcher due to the location it's instantiated.

My issue with relying on this is that 
- I had to read a bunch of code to verify this
- It's refactoring unfriendly (moving the type to instantiate elsewhere will break the guarantee, changes in gRPC factory or bootstraping may break the guarantee and so on).
- It requires a long comment with an explanation why the main thread's dispatcher is the same as the one used by the client so the code is thread safe.

This seems less than ideal from many different angles, maintainability being the most important one, hence I suggested to make this explicit and self explanatory by exposing the client's dispatcher.

This feels natural that when scheduling callbacks and wanting them to run on the same thread as the gRPC client you use, you use its dispatcher. I checked the code got Envoy vs Google gRPC clients and I understand that Google's using a separate thread with final dispatching back, but in both cases it seems possible.
